2021/11/18 Add AirQuality, Historical, and Tropical API

john35452 opened the pull request:
This PR is the combination of following PRs.
AirQuality:
https://github.com/Azure/azure-rest-api-specs/pull/16352
Historical:
https://github.com/Azure/azure-rest-api-specs/pull/16486
Tropical:
https://github.com/Azure/azure-rest-api-specs/pull/15532

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

Commit:
2021/11/18 Add air quality api

Commit:
2021/08/05 Add Tropical api

evaou committed:
2021/11/18 Add Historical api

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
+    "/weather/tropical/storms/active/{format}": {
+      "get": {
+        "description": "**Get Tropical Active Storms**\n\n**Applies to**: Gen 1 (S0 and S1) and Gen 2 pricing tiers.\n\nGet all government-issued active tropical storms. Information about the tropical storms includes, government ID, basin ID, year of origin, name and if it is subtropical.",
+        "operationId": "Weather_GetTropicalActiveStorms",
```
This does not read naturally. `Weather_GetActiveTropicalStorms` makes more sense.

john35452 replied:
@eriklindeman How do you think? This change will affect the title of document page.

eriklindeman replied:
The names of the services should always, be 'Tropical Storm(s)', so I agree with that part, but I think it should be changed to 'Get Tropical Storms Active' to stay uniform with the other three Tropical Storm services if possible:

GetTropicalStormForecast
GetTropicalStormLocations
GetTropicalStormSearch

john35452 replied:
Sure. I have updated the naming to GetTropicalStormsActive.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
+    "/weather/tropical/storms/{format}": {
+      "get": {
+        "description": "**Get Tropical Storm Search**\n\n**Applies to**: Gen 1 (S0 and S1) and Gen 2 pricing tiers.\n\nSearch government-issued tropical storms by year, basin ID, and government ID. Information about the tropical storms includes, government ID, basin ID, status, year, name and if it is subtropical.",
+        "operationId": "Weather_GetTropicalStormSearch",
```
Probably needs an `x-ms-client-name`. Should be something like `SearchTropicalStorm` or `SearchTropicalStormData`

john35452 replied:
Added

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
-            "value": 22,
-            "name": "Snow",
-            "description": "Snow"
+            "name": "windowGeometry",
+            "description": "When true, window geometry data (geoJSON) is included in the response",
```
Why is this not a $ref?  Regardless, as a bool it should be self-documenting with `x-ms-client-name`. Perhaps `include_window_geometry`.

john35452 replied:
Added

john35452 replied:
Maybe I missed something but I only see one windowGeometry in the file. Do you still see two definitions of windowGeometry?

tjprescott replied:
You're right. I only see the one definition.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
-            "value": 37,
-            "name": "HazyMoonlight",
-            "description": "Hazy Moonlight"
+            "name": "current",
+            "description": "When true, return the current storm location",
```
Self-documenting name needed.

john35452 replied:
Added

john35452 replied:
Updated

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
+            "x-ms-enum": {
+              "name": "HourlyDuration",
+              "modelAsString": true,
+              "values": [
+                {
+                  "value": 1,
+                  "description": "1 Hours"
+                },
+                {
+                  "value": 12,
+                  "description": "12 Hours"
+                },
+                {
+                  "value": 24,
+                  "description": "24 Hours"
+                },
+                {
```
Not sure how these will translate to enum properties. Pretty sure they can't just be numbers in most languages.

john35452 replied:
Thanks for the suggestion!
I have added name for all those enums.
Is it ok for you?

tjprescott reviewed (changes requested):
Partial review.

eriklindeman commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
+    },
+    "/weather/tropical/storms/locations/{format}": {
+      "get": {
+        "description": "**Get Tropical Storm Locations**\n\n**Applies to**: Gen 1 (S0 and S1) and Gen 2 pricing tiers.\n\nGet locations of an individual government-issued tropical storms. Information about the tropical storms includes, location coordinates, geometry, basin ID, date, wind details and wind radii.",
```
Change:
Get locations of an individual government-issued tropical storms.

To:
Get location of individual government-issued tropical storms.

john35452 replied:
Updated!

eriklindeman commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +715,1772 @@
+    },
+    "/weather/tropical/storms/forecasts/{format}": {
+      "get": {
+        "description": "**Get Tropical Storm Forecasts**\n\n**Applies to**: Gen 1 (S0 and S1) and Gen 2 pricing tiers.\n\nGet all government-issued tropical storm forecasts. Information about the forecasted tropical storms includes, location, status, date the forecast was created, window, wind speed and wind radii.",
```
Change:
Get all government-issued tropical storm forecasts.

To:
Get individual government-issued tropical storm forecasts.

john35452 replied:
Updated!

eriklindeman commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2801,1226 @@
-      "description": "Description of the weather hazard affecting the trip.",
+    "Window": {
       "type": "object",
+      "description": "Government-issued storm forecast",
```
Change:
Government-issued storm forecast

To:
Contains information about the forecast window for the storm during the specified time period (not the entire cone). If windowGeometry=true in the request, this object will include geoJSON details for window geometry.

eriklindeman reviewed (changes requested):
A couple of minor wording change requests for Forecast and Locations.

Commit:
2021/11/22 Update enum and add x-ms-client-name

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +91,149 @@
+          },
+          {
+            "value": "SP",
+            "description": "SP"
```
Thanks for remind! I have updated them.

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -50,6 +50,14 @@
   "responses": {},
   "parameters": {
+    "Pollutants": {
+      "name": "pollutants",
```
Added

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +91,149 @@
+      "x-ms-parameter-location": "method"
+    },
+    "RequiredWeatherGovernmentId": {
+      "name": "govId",
+      "description": "Government storm Id",
```
Added

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +91,149 @@
+      "x-ms-parameter-location": "method"
+    },
+    "WeatherDetails": {
+      "name": "details",
+      "description": "When true, wind radii summary data is included in the response",
```
Added

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +91,149 @@
+      "x-ms-parameter-location": "method"
+    },
+    "WeatherRadiiGeometry": {
+      "name": "radiiGeometry",
+      "description": "When true, wind radii summary data and geoJSON details are included in the response",
```
Added

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +91,149 @@
       "x-ms-parameter-location": "method"
+    },
+    "WeatherYear": {
+      "name": "year",
+      "description": "Year of the cyclone(s)",
+      "type": "integer",
```
This is the year we want to search for the storms.
It also not a property of the storm, so I think it also not ok to rename to something like StormYear.
So I am ok with keep it as what is it for now.
How do you think?

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +91,149 @@
+      "x-ms-parameter-location": "method"
+    },
+    "RequiredWeatherBasinId": {
+      "name": "basinId",
```
I totally am agree with you, and I tried with this for several time.
But I keep running in to below problems.
An error happens when I add the line to set the parameter to required.
Do you have any suggestion for this?
![image](https://user-images.githubusercontent.com/12066967/142885367-86eb78df-4836-468b-9434-0bda6e3dacd2.png)

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1087,26 +1094,32 @@
                   "description": "12 Hours"
                 },
                 {
+                  "name": "TwentyHours",
```
TwentyFourHours

john35452 replied:
Updated

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -834,6 +839,7 @@
           {
             "name": "windowGeometry",
             "description": "When true, window geometry data (geoJSON) is included in the response",
+            "x-ms-client-name": "include_window_geometry",
```
Camel casing? `includeWindowGeometry`

alankashiwa replied:
Other `x-ms-client-name` should also be camelCase.

john35452 replied:
Updated

Commit:
2021/11/23 Update examples

Commit:
2021/11/23 Update naming

Commit:
2021/11/23 Create parameter for Position

Commit:
2021/11/23 Update active storm

john35452 commented:
@tjprescott 
Please help me to review again, and don't merge it before 12/1.
Really appreciate for your help!

tjprescott commented:
I created a Python APIView of the diff to better comment on the resulting API:
https://apiview.dev/Assemblies/Review/5664808ac3ec49f3ab287f3c9e85f55b?diffRevisionId=d8863f36a80549228fb1df22a3932a76&doc=False&diffOnly=True

tjprescott reviewed (changes requested):
Several more comments direclty in the accompanying APIView:
https://apiview.dev/Assemblies/Conversation/5664808ac3ec49f3ab287f3c9e85f55b

tjprescott commented:
@john35452 I completed a full review of the changes via APIView. There are comments in this review but many more comments in the APIView. Feel free to reach out if you have questions/concerns.

Commit:
2021/11/30 Change reponse to result

Commit:
2021/11/30 Add x-ms-client-name

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +92,167 @@
+    },
+    "Position": {
+      "name": "query",
+      "in": "query",
```
Added

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +643,1703 @@
+        }
+      }
+    },
+    "HourlyForecastResponse": {
```
Thanks for the remind!
I have updated the naming.

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +643,1703 @@
+        }
+      }
+    },
+    "MinuteForecastResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +643,1703 @@
+        }
+      }
+    },
+    "QuarterDayForecastResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +643,1703 @@
       }
     },
-    "HourlyForecast": {
+    "CurrentConditionsResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1214,320 +2361,237 @@
       }
     },
-    "MinuteForecastResponse": {
+    "DailyForecastResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "CurrentConditionsResponse": {
+    "WeatherAlongRouteResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
+        }
+      }
+    },
+    "SevereWeatherAlertsResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "DailyForecastResponse": {
+    "ActiveStormResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "WeatherAlongRouteResponse": {
+    "StormSearchResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "SunGlare": {
+    "StormForecastResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "SevereWeatherAlert": {
+    "LocationResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "SevereWeatherAlertDescription": {
+    "CurrentAirQualityResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
+        }
+      }
+    },
+    "HourlyAirQualityForecastResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
+        }
+      }
+    },
+    "DailyAirQualityForecastResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
       }
     },
-    "DailyIndicesResponse": {
+    "DailyHistoricalActualsResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
+        }
+      }
+    },
+    "DailyHistoricalRecordsResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2660,1226 @@
+        }
+      }
+    },
+    "DailyHistoricalNormalsResponse": {
```
Updated

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -83,6 +92,167 @@
+    },
+    "OptionalWeatherGovernmentId": {
+      "name": "govId",
+      "description": "Government storm Id",
```
Added

john35452 commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -564,621 +643,1703 @@
+      "type": "object",
+      "properties": {
+        "date": {
+          "x-ms-client-name": "dateTime",
```
Updated

Commit:
2021/11/30 Update timestamp, coordinate, and radius naming in SDK

Commit:
2021/11/30 Use WeatherUnit for Concentration

Commit:
2021/11/30 Update Daily Air quality dominantPollutant

Commit:
2021/11/30 Unify response format

Commit:
2021/11/30 Unify CurrentAirQuality and HourlyAirQualityForecast

Commit:
2021/12/01 Update window and weatherValue SDK name

john35452 commented:
@tjprescott Really appreciate for all the comments. I have updated my code. Please help me to review them again.
If they are good enough, please help me to merge since we are going to release this documentation next Monday.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2662,1213 @@
+        }
+      }
+    },
+    "LocationResult": {
```
This should be something like `StormData` or `StormProperties`. Storm location is merely one field among many.

john35452 replied:
This api is like a proxy api for AccuWeather's storm positon api. http://apidev.accuweather.com/developers/tropicalAPIGuide#PositionsGov It not only contains the location info, but some other storm properties. I think it's ok to stick to the naming in AccuWeather to avoid confusion. How do you think?

tjprescott replied:
That API doesn't specify any kind of name, so I'm not sure how not calling it "LocationResponse" would make it confusing. It's already confusing since it contains a whole lot more than location info. Lots and lots of Maps APIs have locations in them, so this is not a good class name. `StormLocation` would be okay here. Since this is a collection item, it should not include Response in the name.

john35452 replied:
Thanks for the suggestion!
I have updated it to StormLocation.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2662,1213 @@
-          "$ref": "#/definitions/HazardIndex"
+        "location": {
+          "$ref": "../../../Common/preview/1.0/common.json#/definitions/LatLongPair",
+          "x-ms-client-name": "coordinate",
```
I think the consistent usage in maps APIs is "coordinates"

john35452 replied:
Updated

tjprescott reviewed (changes requested):
There are also still comments unaddressed in APIView.

Commit:
2021/12/02 Update coordinate and left, right SDK name

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -3316,10 +3316,12 @@
         },
         "right": {
           "$ref": "../../../Common/preview/1.0/common.json#/definitions/LatLongPair",
+          "x-ms-client-name": "BottomRight",
```
Should be camelCase `bottomRight`

john35452 replied:
Updated

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -3316,10 +3316,12 @@
       "properties": {
         "left": {
           "$ref": "../../../Common/preview/1.0/common.json#/definitions/LatLongPair",
+          "x-ms-client-name": "TopLeft",
```
Should be camelCase `topLeft`

john35452 replied:
Updated

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -556,577 +637,1126 @@
+      },
+      "description": "Wind direction"
+    },
+    "WindSpeed": {
+      "type": "object",
+      "description": "Speed of wind in specified unit.",
+      "properties": {
+        "value": {
+          "type": "number",
+          "format": "float",
+          "description": "Rounded value of the speed."
+        },
+        "unit": {
+          "type": "string",
+          "description": "Type of unit for the speed value."
+        },
+        "unitType": {
+          "$ref": "#/definitions/UnitType"
+        }
+      }
+    },
```
This is not referenced by anything and should be removed.

john35452 replied:
WindSpeed is used in windRadiiSummary, and windRadiiSummary is used in StormForecast and StormLocation.

tjprescott replied:
It isn't. They use WeatherUnit. Nothing in the Weather Swagger references this definition (which is just  duplicate of WeatherUnit anyhow).

john35452 replied:
Thanks for the remind! I just removed it.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1564,922 +2662,1215 @@
     },
-    "WeatherAlongRoutePrecipitation": {
-      "description": "Precipitation forecast of the weather along the route.",
+    "StormSearch": {
```
Copied from APIVIew:
Since `StormSearchResult` is taken by the parent object, this should be `StormSearchResultItem`. Unfortunately, `StormSearch` implies that it is some sort of request body.

john35452 replied:
Thanks for the suggestion! I have updated the name.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1249,284 +2194,405 @@
+    "CurrentConditionsResult": {
+      "type": "object",
+      "properties": {
+        "results": {
```
Copied over from APIView:

> **tjprescott**
> You should be specific because the non-tropic storm APIs are being specific.
> 
> **john35452**
> We have a discussion internally, and we would like to use results for all these three new weather apis.

This is not an arbitrary choice you can make just for new APIs. The names need to be consistent with the entire maps effort, which has already decided in the direction of specific collections over "results".

tjprescott replied:
Per our discussion, it's fine to use `x-ms-client-name` to address this. No service change is needed.

john35452 replied:
Sure. I have applied meaningful name in x-ms-client-name in the below commit.
https://github.com/Azure/azure-rest-api-specs/pull/16802/commits/92e55e145299c94d8b87c9e3bf319d3feba6f386

tjprescott reviewed (changes requested):
I've moved all remaining comments over from APIView so there are no open threads there. Once these issues are resolved we can merge.

Commit:
2021/12/03 Add SDK name for results

Commit:
2021/12/03 Improve Object name

Commit:
2021/12/03 Update ActiveStorm

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -3485,6 +3489,7 @@
       "properties": {
         "results": {
           "type": "array",
+          "x-ms-client-name": "airQualityResult",
```
Should be plural.

john35452 replied:
Updated

Commit:
2021/12/03 Remove needless object and update name

Commit:
2021/12/06 Add weather 2.0 directory

Commit:
2021/12/06 Recover Weather 1.0 Swagger file

Commit:
2021/12/07 Remove v2 from operationId

Commit:
2021/12/08 Update readme.md

evaou committed:
Set default API version as 2.0

markweitzel reviewed (commented):
@john35452 A couple of questions on the overall design of the API...

Looking at `/weather/tropical/storms/active/{format}` was there a consideration to have `storms` be the primary collection? So this would look like, `/weather/storms/tropical/active/{format} ? This would allow you to have different kinds of `storms` in the future, e.g. `/weather/storms/cyclones/...` or `/weather/storms/tornadoes/...` 

Following this pattern may also simplify the shape of the other APIs. For example...
`/weather/storms/tropical/forecasts/{}`. 

Also... it seems like there is the concept of a `stormId`, e.g. `goveId`. Would you ever have anything like... `/weather/storms/{govId}/forecast` ? This might be more simple and straight forward for developers to consume. 

I've got a similar question on `/weather/severe/alerts/{format}` -- This feels like `alerts` should be the primary collection, e.g. `/weather/alerts/severe/...` (This would also be similar to how [weather.gov](https://www.weather.gov/documentation/services-web-api) models their alerts. 

Lastly, you put this file in `stable/2.0` -- This is likely something you want to preview first, so I'd consider moving it to `\preview\preview-1.1`

markweitzel reviewed (changes requested):
## Here are the [comments from the API Stewardship Review Board](https://github.com/Azure/azure-rest-api-specs/issues/16993#issuecomment-994118898)

MushMal commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+    }
+  },
+  "paths": {
+    "/weather/forecast/hourly/{format}": {
```
Q: Could the temporal aspect (hourly, minute, quarter daily, daily, current) be an extensible enum parameter rather a path?

john35452 replied:
Thanks for the suggestion!
This is not the new weather api which is out of scope of this time.

MushMal commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+        }
+      }
+    },
+    "/weather/airQuality/forecast/daily/{format}": {
```
Similar question here - could hourly, daily, etc.. be extensible enum parameters?

john35452 replied:
The response formats of daily and hourly forecase are slightly different.
So I think we might need to separate them.

MushMal reviewed (commented):
We discussed the pagination considerations for some of the APIs and the response sizes earlier. I have a comment about the temporal sub-path parameterization but other than that looks good. I also see a lot of semantic errors loading the file in the swagger editor. Are these benign?

```
Semantic error at paths./weather/forecast/hourly/{format}
Declared path parameter "format" needs to be defined as a path parameter at either the path or operation level
```

mikekistler commented on custom-words.txt:
```diff
@@ -2456,4 +2458,5 @@ testrun
-servermetrics
\ No newline at end of file
+servermetrics
+Obuscated
```
I don't believe "Obuscated" is a word.  You probably want "Obfuscated".

john35452 replied:
Good catch! 
I also notice this is not used any more, so I remove this term from the custom-words list.

mikekistler reviewed (commented):
I left a number of questions/suggestions that I hope you will give proper consideration.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+      }
+    },
+    "WeatherWaypoint": {
+      "x-ms-client-name": "WaypointForecast",
```
Thanks for the suggestion! But this is not one of the new weather api.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+            "name": "duration",
+            "description": "Specifies for how many days the daily forecast responses are returned. Available values are\n  * `1` - Return forecast data for the next day. Returned by default.\n  * `5` - Return forecast data for the next 5 days.\n  * `10` - Return forecast data for the next 10 days.\n  * `25` - Return forecast data for the next 25 days. Only available in S1 SKU.\n  * `45` - Return  ...
+            "in": "query",
+            "type": "integer"
```
Thanks for the suggestion!
This is not the new weather api which is out of scope.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+            "name": "details",
+            "description": "Return full details for the severe weather alerts. Available values are\n  * `true` - Returns full details. By default all details are returned.\n  * `false` - Returns a truncated version of the alerts data, which excludes the area-specific full description of alert details (`alertDetails`).",
+            "in": "query",
+            "type": "string"
```
Thanks for the suggestion!
This is not the new weather api which is out of scope of this time.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+            "name": "duration",
+            "description": "Specifies for how many days the daily indices are returned. By default, the indices data for the current day will be returned. When requesting future indices data, the current day is included in the response as day 1. Available values are\n  * `1` - Return daily index data for the current day. Default value.\n  * `5` - Return 5 days of daily index data starting from th ...
+            "in": "query",
+            "type": "integer"
```
Thanks for the suggestion!
This is not the new weather api which is out of scope of this time.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+      "type": "string",
+      "in": "query",
+      "required": true,
+      "default": "2.0",
```
Thanks a lot! I have removed the default value here.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+            "description": "Return full details for the current conditions. Available values are\n  * `true` - Returns full details. By default all details are returned.\n  * `false` - Returns a truncated version of the current condition data, which includes observation date time, weather phrase, icon code, precipitation indicator flag, and temperature.",
+            "in": "query",
+            "type": "string"
+          },
```
Thanks for the suggestion!
This is not the new weather api which is out of scope of this time.

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+            "type": "integer"
+          },
+          {
+            "$ref": "../../../Common/preview/1.0/common.json#/parameters/Language"
```
Yes. Ideally it should be like that.
However, currently there is no stable version of common library and this common library is used in other stable api documentation as well.

mikekistler replied:
There are at least two ways to deal with this:
1. Copy the schemas you need from the Common file into your API, or
2. Create a "stable" version of the Common file.

Leaving it as is is highly not recommended.

evaou replied:
After consulting SDK members, they will move the related files to stable version once the SDKs are launched.

Commit:
2022/01/07 Remove default value for api-version

Commit:
2022/01/07 Remove needless custom-words

john35452 commented on specification/maps/data-plane/Weather/stable/2.0/weather.json:
```diff
@@ -0,0 +1,3869 @@
+    "OptionalWeatherGovernmentId": {
+      "name": "govId",
+      "description": "Government storm Id",
+      "x-ms-client-name": "governmentStormId",
```
WeatherGovernmentId is required for some apis and some are optional.
That's why we have two different parameters here.

mikekistler replied:
Ah yes ... I see that @tjprescott also commented on this. And it is very unfortunate but I agree that OpenAPI 2.0 gives you no way to reuse a definition for this query parameter and just switch the `required` value.  This _is_ possible in OpenAPI 3.0, but ... sigh.

Commit:
2022/01/11 Update api-version and add nextLink

Commit:
2022/01/11 Update api-version in readme.md

tjprescott commented on specification/maps/data-plane/Weather/stable/1.1/examples/GetActiveStorms.json:
```diff
@@ -1,6 +1,6 @@
 {
   "parameters": {
-    "api-version": "2.0",
+    "api-version": "1.1",
```
I thought the discussion at the meeting was for date-based versions since that's what other services with an API version query parameter do?

evaou replied:
Per the discussion and conclusion in the 2nd weather api review session, we would apply api version 1.1 and I've shared the related recording with Travis.

mikekistler reviewed (approved):
Looks good.  Thanks for addressing my comments.

tjprescott commented:
@markwietzel have the comments from #16993 been addressed?

john35452 commented:
> 

Hi Mark,

Thanks for noting the conclusion of discussion here.
Here comes the reply or status for the question you mentioned here.
Different type of storm: 
As the discussion in the meeting, we decide to follow the AccuWeather to keep the term tropical storm.

API path pattern:
We also have unified those patterns like airquality/hourly/forecast -> airquality/hourly/forecasts.

Severe alert:
This api is not one of our new weather api.

API-version:
We are going to GA the new api, and we have change the new version from 2.0 to 1.1.
So I moved new files under stable/1.1 directory.

Really appreciate for your review comment!
We have updated our PR.
Please help us to review again.

john35452 commented:
> @john35452 A couple of questions on the overall design of the API...
> 
> Looking at `/weather/tropical/storms/active/{format}` was there a consideration to have `storms` be the primary collection? So this would look like, `/weather/storms/tropical/active/{format} ? This would allow you to have different kinds of `storms`in the future, e.g.`/weather/storms/cyclones/...`or`/weather/storms/tornadoes/...`
> 
> Following this pattern may also simplify the shape of the other APIs. For example... `/weather/storms/tropical/forecasts/{}`.
> 
> Also... it seems like there is the concept of a `stormId`, e.g. `goveId`. Would you ever have anything like... `/weather/storms/{govId}/forecast` ? This might be more simple and straight forward for developers to consume.
> 
> I've got a similar question on `/weather/severe/alerts/{format}` -- This feels like `alerts` should be the primary collection, e.g. `/weather/alerts/severe/...` (This would also be similar to how [weather.gov](https://www.weather.gov/documentation/services-web-api) models their alerts.
> 
> Lastly, you put this file in `stable/2.0` -- This is likely something you want to preview first, so I'd consider moving it to `\preview\preview-1.1`

> 

Hi Mark,

Thanks for noting the conclusion of discussion here.
Here comes the reply or status for the question you mentioned here.
Different type of storm: 
As the discussion in the meeting, we decide to follow the AccuWeather to keep the term tropical storm.

API path pattern:
We also have unified those patterns like airquality/hourly/forecast -> airquality/hourly/forecasts.

Severe alert:
This api is not one of our new weather api.

API-version:
We are going to GA the new api, and we have change the new version from 2.0 to 1.1.
So I moved new files under stable/1.1 directory.

Really appreciate for your review comment!
We have updated our PR.
Please help us to review again.

markweitzel reviewed (approved):
Thank you for updating this and incorporating the changes.
